crates/rooch/src/extended_checks/metadata.rs

```diff
@@ -0,0 +1,280 @@
+
+                    if private_generics_types.is_some() {
+                        let private_generics_types_indices = private_generics_types.unwrap();
+
+                        for generic_type_index in private_generics_types_indices {
+                            let type_arg = type_arguments.get(*generic_type_index).unwrap();
+                            let (defined_in_current_module, struct_name) =
+                                is_defined_in_current_module(&view, type_arg);
+
+                            if !defined_in_current_module {
+                                panic!(
+                                    "{}",
+                                    format!(
+                                        "resource struct {:?} not defined in current module",
+                                        struct_name
+                                    )
+                                );
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        for (private_generics_func_name, types_list) in type_name_indices {
+            let type_params_idicies = self
+                .output
+                .entry(module.self_id().clone())
+                .or_default()
+                .private_generics_indices
+                .entry(private_generics_func_name)
+                .or_default();
+
+            let _ = types_list
+                .iter()
+                .map(|index| type_params_idicies.push(*index))
+                .collect::<Vec<_>>();
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------------
+// Helpers
+
+impl<'a> ExtendedChecker<'a> {
+    fn has_attribute(&self, fun: &FunctionEnv, attr_name: &str) -> bool {
+        fun.get_attributes().iter().any(|attr| {
+            if let Attribute::Apply(_, name, _) = attr {
+                self.env.symbol_pool().string(*name).as_str() == attr_name
+            } else {
+                false
+            }
+        })
+    }
+}
+
+fn is_defined_in_current_module(
+    view: &BinaryIndexedView,
+    type_arg: &SignatureToken,
+) -> (bool, String) {
+    match type_arg {
+        SignatureToken::Struct(idx) | SignatureToken::StructInstantiation(idx, _) => {
+            let shandle = view.struct_handle_at(*idx);
+            (
+                view.self_handle_idx() == Some(shandle.module),
+                view.identifier_at(shandle.name).to_string(),
+            )
+        }
+        SignatureToken::TypeParameter(_)
+        | SignatureToken::Bool
+        | SignatureToken::U8
+        | SignatureToken::U16
+        | SignatureToken::U32
+        | SignatureToken::U64
+        | SignatureToken::U128
+        | SignatureToken::U256
+        | SignatureToken::Address
+        | SignatureToken::Vector(_)
+        | SignatureToken::Signer
+        | SignatureToken::Reference(_)
```

Comment:
Do we not support `Reference`? Is the code as follows worked?

```
#[private_generics(T)]
public fun create_box<T>(value: &T): Box<T1> {
    Box<T1> { value: value }
}
```

Reply:
I have tested the following code and it works properly:

```move
#[private_generics(T1, T3)]
public fun create_box<T1: store+copy, T2, T3>(value: &T1): Box<T1> {
    Box<T1> { value: *value }
} 

public fun run() {
    let data = &Data1{ v: 123 };
    let box_val = create_box<Data1, Data2, Box<u32>>(data);
    let _ = box_value(&box_val);
}   
```

Because we are checking the generic parameter `Data1` (which is an ordinary struct) passed when calling `create_box<Data1, Data2, Box<u32>>(data);` and not the type of the `value: &T1` parameter in the create_box function (which is a reference type), the code works properly.
Unless we not only check the type of the generic parameter, but also the type of the actual parameter, the code will not work properly.